src/validate.ts

```diff
@@ -39,33 +44,57 @@ function responseValidation(apiResponse: object, schema: object) {
  * @param apiResponse User defined api response.
  * @return Errors from AJV validation, if any. Undefined otherwise.
  */
-export function validate(intentRequest: object, apiResponse: object) {
+export function validate(intentRequest: object, apiResponse: object, syncData?: object){
   const responseType = intentRequest['inputs'][0]['intent'];
   if (responseType === 'action.devices.SYNC') {
-    // validate traits with attributes schema
+    const validateSyncAPI = responseValidation(apiResponse, SYNC_RESPONSE_SCHEMA); 
     const syncErrors : object[] = [];
-    const syncDevices = apiResponse['payload']['devices'];
-    const syncDevicesLength = syncDevices.length;
-    for (let i = 0; i < syncDevicesLength; i++) {
-      const traits = syncDevices[i]['traits'];
-      const traitsLength = syncDevices[i]['traits'].length;
-      const attributes = syncDevices[i]['attributes'];
-      if (attributes === undefined) {
-        return responseValidation(apiResponse, SYNC_RESPONSE_SCHEMA);
-      }
-      for (let j = 0; j < traitsLength; j++) {
-        const trait = traits[j];
-        if (trait in TRAIT_ATTRIBUTES_EXPECT) {
-          const validateTraitRes = responseValidation(attributes, TRAIT_ATTRIBUTES_EXPECT[trait]);
-          if (validateTraitRes) {
-            return syncErrors.push(...validateTraitRes);
+    if (validateSyncAPI === undefined){
+        const syncDevices = apiResponse['payload']['devices'];
+        const syncDevicesLength = syncDevices.length;
+        for (let i = 0; i < syncDevicesLength; i++) {
+          const traits = syncDevices[i]['traits'];
+          const traitsLength = syncDevices[i]['traits'].length;
+          const attributes = syncDevices[i]['attributes'];
+          if (attributes === undefined) {
+            return validateSyncAPI;
```

Comment:
I don't think this is true? This would bail out on the first device that is missing attributes? A lack of attributes doesn't mean it shouldn't be validated against the trait schemas, it just means that the **content** to be validated is an empty object `{}`

So maybe this just needs to be `const attributes = syncDevices[i]['attributes'] || {}` and drop the "if"?

Reply:
The logic for this statement was that if the attributes don't exist you would just need to look at the whole schema instead. I got to this point because when this statement is removed, an error occurs in the intent handler SYNC test, where the response received is  [{"dataPath": "", "keyword": "type", "message": "should be object", "params": {"type": "object"}, "schemaPath": "#/type"}] instead of undefined.

Reply:
Right, because you wouldn't be able to pass "undefined" as a parameter to your attributes validator below (and you shouldn't), but recall that we discussed validating "no attributes" is just as important as those that are present. If attributes are empty but the trait requires one, we can't skip that case simply because no attributes are there. We just have to pass a representation that the validator can work with, which is an empty object (i.e. `{}`)

This isn't the same as validating the rest of the schema (which you are now doing earlier).

Reply:
Got it, that makes sense. I'll go ahead and commit these changes! Tried it out again with the || {} portion and it works as intended.